mapping_parenting_tech/analysis/prototyping/clustering_descriptions.py

```diff
@@ -57,45 +76,380 @@
+
+neigh.fit(X_list, y_list)
+
+for x in unassigned_indices:
+    s_df.iat[x, s_cluster_col] = neigh.predict([s_embedding_clustering[x]])
+
+print(f"{len(s_df[s_df.s_cluster == -1].s_cluster.to_list())} unassigned apps remain.")
+
+# %%
+# Visualise using altair (NB: -1=points haven't been assigned to a cluster)
+s_fig = (
+    alt.Chart(s_df.reset_index(), width=750, height=750)
+    .mark_circle(size=60)
+    .encode(
+        x="x", y="y", tooltip=["s_cluster", "appId", "summary"], color="s_cluster:N"
+    )
+).interactive()
+
+s_fig
+
+# %%
+s_relevant_clusters = set()
+for a_id in set(extra_apps + apps_of_interest):
+    if a_id in s_df.appId.values:
+        s_relevant_clusters.update(s_df[s_df.appId == a_id].s_cluster.values)
+        if s_df[s_df.appId == a_id].s_cluster.values == -1:
+            print(f"{a_id} not in cluster")
+
+s_relevant_clusters
+
+# %%
+s_cluster_options = {
+    "relevant": {
+        2: "stories/lullabies/music",
+        4: "learning through play / learning to code/ professional development",
+        6: "health / medical (adult)",
+        8: "animal-related (identify, sounds etc.)",
+        9: "play / make-belief",
+        10: "sleep (helping babies to)",
+        11: "education / parenting support",
+        17: "pregnancy",
+        18: "baby monitoring / development tracking (photos)",
+        19: "number games",
+        24: "learning games for toddlers",
+        26: "learning games (mostly ABCs)",
+        27: "learning ABCs",
+        28: "learning English (inc. as a foreign language)",
+    },
+    "interesting": {
+        1: "drawing / colouring",
+        12: "period / fertility tracking",
+        25: "learning games",
+    },
+    "unlikely": {
+        6: "student learning",
+        16: "meditation / calm / sleep (for adults)",
+        23: "mental health",
+    },
+    "discard": {
+        0: "Setting light levels (eg screen brightness)",
+        3: "driving test",
+        5: "engineering",
+        7: "connecting with people",
+        13: "personal organisation",
+        14: "child safety / web filtering",
+        15: "outdoor / mapping",
+        20: "religion",
+        21: "minfulness / Buddhism",
+        22: "motivation",
+    },
+}
+
+
+# %%
+def get_cluster_info(needle: int, haystack: dict):
+    """
+    Get information about a cluster from the 'relevance' dictionaries based on a cluster's id
+
+    Example usage: relevance, purpose = get_cluster_info(4, s_cluster_options)
+        Returns: relevance = "relevant"; purpose = "learning through play / learning to code/ professional development"
+
+    Args:
+        needle, int: the cluster to search for
+        haystack, dict: the dict with the cluster info, within which you're looking
+
+    Returns:
+        Tuple(the key of the cluster that you're looking for, the purpose of the cluster)
+    """
+
+    for k, v in haystack.items():
+        if needle in v.keys():
+            return (k, v[needle])
+
+    return 0
```

Comment:
This function still threw an error for me (the same `TypeError: cannot unpack non-iterable int object`).

The following fixed it; you could probably also do `return (None, None)` instead of `(0,0)`
```suggestion
    return (0, 0)
```